pydicom/pixel_data_handlers/util.py

```diff
@@ -1,14 +1,246 @@
+    if nr_entries < 0:
+        nr_entries += 2**16
+
+    # May be negative if Pixel Representation is 1
+    first_map = lut_desc[1]
+    # Actual bit depth may be larger (8 bit entries in 16 bits allocated)
+    nominal_depth = lut_desc[2]
+    dtype = np.dtype('uint{:.0f}'.format(nominal_depth))
+
+    luts = []
+    if 'RedPaletteColorLookupTableData' in ds:
+        # LUT Data is described by PS3.3, C.7.6.3.1.6
+        r_lut = ds.RedPaletteColorLookupTableData
+        g_lut = ds.GreenPaletteColorLookupTableData
+        b_lut = ds.BluePaletteColorLookupTableData
+        a_lut = getattr(ds, 'AlphaPaletteColorLookupTableData', None)
+
+        actual_depth = len(r_lut) / nr_entries * 8
+        dtype = np.dtype('uint{:.0f}'.format(actual_depth))
+
+        for lut in [ii for ii in [r_lut, g_lut, b_lut, a_lut] if ii]:
+            luts.append(np.frombuffer(lut, dtype=dtype))
+    elif 'SegmentedRedPaletteColorLookupTableData' in ds:
+        # Segmented LUT Data is described by PS3.3, C.7.9.2
+        r_lut = ds.SegmentedRedPaletteColorLookupTableData
+        g_lut = ds.SegmentedGreenPaletteColorLookupTableData
+        b_lut = ds.SegmentedBluePaletteColorLookupTableData
+        a_lut = getattr(ds, 'SegmentedAlphaPaletteColorLookupTableData', None)
+
+        endianness = '<' if ds.is_little_endian else '>'
+        byte_depth = nominal_depth // 8
+        fmt = 'B' if byte_depth == 1 else 'H'
+        actual_depth = nominal_depth
+
+        for seg in [ii for ii in [r_lut, g_lut, b_lut, a_lut] if ii]:
+            len_seg = len(seg) // byte_depth
+            s_fmt = endianness + str(len_seg) + fmt
+            lut = _expand_segmented_lut(unpack(s_fmt, seg), s_fmt)
+            luts.append(np.asarray(lut, dtype=dtype))
+    else:
+        raise ValueError("No suitable Palette Color Lookup Table Module found")
+
+    if actual_depth not in [8, 16]:
+        raise ValueError(
+            "The bit depth of the LUT data '{:.1f}' is invalid (only 8 or 16 "
+            "bits per entry allowed)".format(actual_depth)
+        )
+
+    if False in [len(item) == len(luts[0]) for item in luts]:
+        raise ValueError("LUT data must be the same length")
+
+    # IVs < `first_map` get set to first LUT entry (i.e. 0)
+    clipped_iv = np.zeros(arr.shape, dtype=dtype)
+    # IVs >= `first_map` are mapped by the Palette Color LUTs
+    # `first_map` may be negative, positive or 0
+    mapped_pixels = arr >= first_map
+    clipped_iv[mapped_pixels] = arr[mapped_pixels] - first_map
+    # IVs > number of entries get set to last entry
+    np.clip(clipped_iv, 0, nr_entries - 1, out=clipped_iv)
+
+    # Output array may be RGB or RGBA
+    out = np.empty(list(arr.shape) + [len(luts)], dtype=dtype)
+    for ii, lut in enumerate(luts):
+        out[..., ii] = lut[clipped_iv]
+
+    return out
+
+
+def apply_modality_lut(arr, ds):
+    """Apply a modality lookup table or rescale operation to `arr`.
```

Comment:
probably a function for a VOI LUT sequence can be added later if needed... should be very similar

Reply:
Yeah, I didn't want to do too much in this PR